pkg/daemon/utils/ip_range.go

```diff
@@ -0,0 +1,226 @@
+)
+
+type IPRange struct {
+	start net.IP
+	end   net.IP
+}
+
+func CreateIPRange(start, end net.IP) (*IPRange, error) {
+	if start == nil || end == nil {
+		return nil, fmt.Errorf("start and end should not be nil")
+	}
+
+	if ip.Cmp(start, end) > 0 {
+		return nil, nil
+	}
+
+	return &IPRange{
+		start: start,
+		end:   end,
+	}, nil
+}
+
+func (ir *IPRange) TryAddIP(ipAddr net.IP) (success bool) {
+	if ipAddr.Equal(ip.PrevIP(ir.start)) {
+		ir.start = ipAddr
+		return true
+	} else if ipAddr.Equal(ip.NextIP(ir.end)) {
+		ir.end = ipAddr
+		return true
+	} else if ip.Cmp(ipAddr, ir.start) >= 0 && ip.Cmp(ipAddr, ir.end) <= 0 {
+		// a range exist which includes this ip address
+		return true
+	}
+	return false
+}
+
+// Translate a subnet range into a series ip block description.
+func FindSubnetExcludeIPBlocks(cidr *net.IPNet, includedRanges []*IPRange, gateway net.IP,
+	excludeIPs, reservedIPs []net.IP) ([]*net.IPNet, error) {
+
+	cidrStart := cidr.IP
+	cidrEnd := LastIP(cidr)
+
+	var excludeIPRanges []*IPRange
+
+	sort.Slice(includedRanges, func(i, j int) bool {
+		return ip.Cmp(includedRanges[i].start, includedRanges[j].start) < 0
+	})
+
+	for currentIPRangeIndex, currentIPRange := range includedRanges {
+		if ip.Cmp(currentIPRange.start, cidrStart) < 0 || ip.Cmp(currentIPRange.end, cidrEnd) > 0 {
+			return nil, fmt.Errorf("ip range %v~%v is out of cidr %v",
+				currentIPRange.start, currentIPRange.end, cidr)
+		}
+
+		if currentIPRangeIndex < (len(includedRanges)-1) &&
+			ip.Cmp(currentIPRange.end, includedRanges[currentIPRangeIndex+1].start) >= 0 {
+			return nil, fmt.Errorf("ip range is overlapped for range %v~%v and %v~%v",
+				currentIPRange.start, currentIPRange.end,
+				includedRanges[currentIPRangeIndex+1].start, includedRanges[currentIPRangeIndex+1].end)
+		}
+
+		if currentIPRangeIndex == 0 {
+			// add [cidrStart, currentRangeStartPrev] to exclude ip ranges
+			currentRangeStartPrev := ip.PrevIP(currentIPRange.start)
+
+			ipRange, err := CreateIPRange(cidrStart, currentRangeStartPrev)
+			if err != nil {
+				return nil, fmt.Errorf("create ip range for %v~%v failed: %v", cidrStart, currentRangeStartPrev, err)
+			}
+
+			if ipRange != nil {
+				excludeIPRanges = append(excludeIPRanges, ipRange)
+			}
+		}
+
+		var nextRangeStartPrev net.IP
+		if currentIPRangeIndex == len(includedRanges)-1 {
+			nextRangeStartPrev = cidrEnd
+		} else {
+			nextRangeStartPrev = ip.PrevIP(includedRanges[currentIPRangeIndex+1].start)
+		}
+
+		// add [endNext, nextRangeStartPrev] to exclude ip ranges
+		endNext := ip.NextIP(currentIPRange.end)
+
+		ipRange, err := CreateIPRange(endNext, nextRangeStartPrev)
+		if err != nil {
+			return nil, fmt.Errorf("create ip range for %v~%v failed: %v", endNext, nextRangeStartPrev, err)
+		}
+
+		if ipRange != nil {
+			excludeIPRanges = append(excludeIPRanges, ipRange)
+		}
+	}
+
+	allExcludedIPs := excludeIPs
+	if len(reservedIPs) != 0 {
+		allExcludedIPs = append(allExcludedIPs, reservedIPs...)
```

Comment:
reservedIPs should be kept not excluded

Reply:
ReservedIPs can not be allocated to pod either, and also can be used to "exclude" ip, so I think is better to keep it.

Reply:
fixed